File: matlab/vargplvmExtractParam.m
function [params, names] = vargplvmExtractParam(model)

% VARGPLVMEXTRACTPARAM Extract a parameter vector from a variational GP-LVM model.
% FORMAT
% DESC extracts a parameter vector from a given VARGPLVM structure.
% ARG model : the model from which parameters are to be extracted.
% RETURN params : the parameter vector extracted from the model.
%
% DESC does the same as above, but also returns parameter names.
% ARG model : the model structure containing the information about
% the model.
% RETURN params : a vector of parameters from the model.
% RETURN names : cell array of parameter names.
%
% COPYRIGHT : Michalis K. Titsias and Neil D. Lawrence, 2009-2011
%
% Modifications: Andreas C. Damianou, 2010-2011 
%
% SEEALSO : vargplvmCreate, vargplvmExpandParam, modelExtractParam

% VARGPLVM

%%% Parameters must be returned as a vector in the following order (left to right) 
% - parameter{size} -
% vardistParams{model.vardist.nParams} % mu, S
%       OR
% [dynamicsVardistParams{dynamics.vardist.nParams} dynamics.kernParams{dynamics.kern.nParams}] % mu_bar, lambda
% inducingInputs{model.q*model.k}
% kernelParams{model.kern.nParams}
% beta{prod(size(model.beta))}


if nargout > 1
  returnNames = true;
else
  returnNames = false;
end 

if isfield(model, 'dynamics') & ~isempty(model.dynamics)
    % [VariationalParameters(reparam)   dynKernelParameters]
    if returnNames
        [dynParams, dynParamNames] = modelExtractParam(model.dynamics);
        names = dynParamNames;
    else
        dynParams = modelExtractParam(model.dynamics);
    end
    params = dynParams;
else
    % Variational parameters 
    if returnNames
        %[varParams, varNames] = vardistExtractParam(model.vardist);
        [varParams, varNames] = modelExtractParam(model.vardist);
        %names = varNames{:}; %%% ORIGINAL
        names = varNames; %%% NEW
    else
        %varParams = vardistExtractParam(model.vardist);
        varParams = modelExtractParam(model.vardist);
    end
    params = varParams;
end


% Inducing inputs: They are included as separate parameters except if:
% a) we fix the inducing points to the variational means 
% b) we fix them to some initial values (not necessarily same as
% variational means)
if ~model.fixInducing
    if ~isfield(model, 'learnInducing') || (isfield(model, 'learnInducing') && model.learnInducing)
        params =  [params model.X_u(:)'];
        if returnNames
            for i = 1:size(model.X_u, 1)
                for j = 1:size(model.X_u, 2)
                    X_uNames{i, j} = ['X_u(' num2str(i) ', ' num2str(j) ')'];
                end
            end
            names = {names{:}, X_uNames{:}};
        end
    end
end


% Kernel parameters  
if returnNames
  [kernParams, kernParamNames] = kernExtractParam(model.kern); 
  for i = 1:length(kernParamNames)
    kernParamNames{i} = ['Kernel, ' kernParamNames{i}];
  end
  names = {names{:}, kernParamNames{:}};
else
  kernParams = kernExtractParam(model.kern);
end
params = [params kernParams];


% beta in the likelihood 
if model.optimiseBeta
   
   if ~isstruct(model.betaTransform)
       fhandle = str2func([model.betaTransform 'Transform']);
       betaParam = fhandle(model.beta, 'xtoa');
   else
      if isfield(model.betaTransform,'transformsettings') && ~isempty(model.betaTransform.transformsettings)
          fhandle = str2func([model.betaTransform.type 'Transform']);
          betaParam = fhandle(model.beta, 'xtoa', model.betaTransform.transformsettings);
      else
          error('vargplvmExtractParam: Invalid transform specified for beta.'); 
      end
   end   

   params = [params betaParam(:)'];
   
   if returnNames
     for i = 1:length(betaParam)
       betaParamNames{i} = ['Beta ' num2str(i)];
     end
     names = {names{:}, betaParamNames{:}};
   end
end

File: vargplvmR/R/vargplvmExtractParam.R
vargplvmExtractParam <-
function (model, only.values = TRUE, untransformed.values = FALSE)
{
# % VARGPLVMEXTRACTPARAM Extract a parameter vector from a variational GP-LVM model.
# % FORMAT
# % DESC extracts a parameter vector from a given VARGPLVM structure.
# % ARG model : the model from which parameters are to be extracted.
# % RETURN params : the parameter vector extracted from the model.
# %
# % DESC does the same as above, but also returns parameter names.
# % ARG model2 : the model structure containing the information about
# % the model.
# % RETURN params : a vector of parameters from the model.
# % RETURN names : cell array of parameter names.
# %
# % COPYRIGHT : Michalis K. Titsias and Neil D. Lawrence, 2009-2011
# %
# % Modifications: Andreas C. Damianou, 2010-2011 
# %
# % SEEALSO : vargplvmCreate, vargplvmExpandParam, modelExtractParam
# 
# % VARGPLVM

# % Parameters must be returned as a vector in the following order (left to right) 
# % - parameter{size} -
# % vardistParams{model.vardist.nParams} % mu, S
# %       OR
# % [dynamicsVardistParams{dynamics.vardist.nParams} dynamics.kernParams{dynamics.kern.nParams}] % mu_bar, lambda
# % inducingInputs{model.q*model.k}
# % kernelParams{model.kern.nParams}
# % beta{prod(size(model.beta))}


if (!only.values)
returnNames <- TRUE 
else
returnNames <- FALSE 


if (("dynamics" %in% names(model)) && !is.null(model$dynamics))
{
  cat("vargplvmExtractParam to do ..\n")
# % [VariationalParameters(reparam)   dynKernelParameters]
  #   To do
  # # if (returnNames)
  # [dynParams, dynParamNames] <- modelExtractParam(model.dynamics) 
  # names <- dynParamNames 
  # else
  #   dynParams <- modelExtractParam(model.dynamics) 
  # end
  # params <- dynParams 
} else {
# % Variational parameters 
  if (returnNames)
  {
    cat("to do vargplvmExtractParam\n")
# %[varParams, varNames] <- vardistExtractParam(model.vardist) 
    # [varParams, varNames] <- modelExtractParam(model$vardist) 
# %names <- varNames{:}  %%% ORIGINAL
    # names <- varNames  %%% NEW
  }
  else {
# %varParams <- vardistExtractParam(model$vardist) 
    varParams <- modelExtractParam(model$vardist) 
  }
params <- varParams 
}


# % Inducing inputs 
if (!model$fixInducing)
{
  params <- c(params, c(matrix(model$X_u, 1, prod(dim(model$X_u)))))
  if (returnNames) 
  {
    for (i in 1:dim(model$X_u)[1])
    {
      for (j in 1:dim(model$X_u)[2])
        X_uNames[i, j] <- paste("X_u(", i, ",", j, ")", sep = "") 
    }
    names <- c(matrix(names, ncol = 1) , matrix(X_uNames, ncol = 1)) 
  } 
}
#     % Kernel parameters  
#     if returnNames
#     [kernParams, kernParamNames] <- kernExtractParam(model$kern)  
#     for i <- 1:length(kernParamNames)
#     kernParamNames{i} <- ['Kernel, ' kernParamNames{i}] 
#     end
#     names <- {names{:}, kernParamNames{:}} 
#     else
      kernParams <- kernExtractParam(model$kern, matlabway = TRUE) 
#     end
    params <- c(params, kernParams) 
    
    
#     % beta in the likelihood 
if (model$optimiseBeta)
{
    if (!is.list(model$betaTransform))
    {
      fhandle <- paste(model$betaTransform, "Transform", sep = "") 
      betaParam <- do.call(fhandle, list(model$beta, "xtoa", matlabway = TRUE)) 
    } else {
      if (("transformsettings" %in% names(model$betaTransform)) && (length(model$betaTransform$transformsettings)>0))
      {
        fhandle <- paste(model$betaTransform$type, "Transform", sep = "") 
        betaParam <- do.call(fhandle, list(model$beta, "xtoa", model$betaTransform$transformsettings)) 
      } else {
        stop("vargplvmExtractParam: Invalid transform specified for beta.\n")
      }
    }    
    params <- c(params, matrix(betaParam, nrow = 1))
   
    if (returnNames)
    {
      betaParamNames <- paste("Beta", 1:length(betaParam), sep="") 
      names(params) <- c(names, betaParamNames)
    }
}
return (params)
}
